library/core/src/num/mod.rs

```diff
@@ -846,37 +849,79 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+    if mem::size_of::<T>() <= mem::size_of::<u32>() && radix == 10 {
+        // The ALU can reorder these adds and do more in parallel
+        // as each mul isn't dependent on the previous answer.
+        let factors = &MULTIPLIER[MULTIPLIER.len() - digits.len()..];
+        let mut idx = 0;
+        if is_positive {
+            // The number is positive
+            for &c in digits {
+                let x = match (c as char).to_digit(radix) {
+                    Some(x) => x,
+                    None => return Err(PIE { kind: InvalidDigit }),
+                };
+                let x = match factors[idx].checked_mul(x) {
+                    Some(result) => result,
+                    None => return Err(PIE { kind: PosOverflow }),
+                };
+                result = match result.checked_add(x) {
+                    Some(result) => result,
+                    None => return Err(PIE { kind: PosOverflow }),
+                };
+                idx += 1;
+            }
+        } else {
+            // The number is negative
+            for &c in digits {
+                let x = match (c as char).to_digit(radix) {
+                    Some(x) => x,
+                    None => return Err(PIE { kind: InvalidDigit }),
+                };
+                let x = match factors[idx].checked_mul(x) {
+                    Some(result) => result,
+                    None => return Err(PIE { kind: NegOverflow }),
+                };
+                result = match result.checked_sub(x) {
+                    Some(result) => result,
+                    None => return Err(PIE { kind: NegOverflow }),
+                };
+                idx += 1;
+            }
```

Comment:
Is it necessary to have so many duplicated code here? Also, why not put the extra condition branch in the error flow rather the default flow?

```suggestion
        for &c in digits {
            let x = match (c as char).to_digit(radix) {
                Some(x) => x,
                None => return Err(PIE { kind: InvalidDigit }),
            };
            let x = match factors[idx].checked_mul(x) {
                Some(result) => result,
                None => return Err(PIE { kind: if is_positive { PosOverflow } else { NegOverflow } }),
            };
            result = match result.checked_add(x) {
                Some(result) => result,
                None => return Err(PIE { kind: if is_positive { PosOverflow } else { NegOverflow } }),
            };
            idx += 1;
```

Reply:
I think you missed there was a `checked_sub` in there? Having an if in the loop to choose between checked_add and checked_sub kills performance - I can attest to that! I did wonder if we could do better too. You can't add and then negate as the negative ranges are slightly bigger. So you could always subtract and then !+1 to make positive, which would work great for signed types (with a check to see if val == T::Min) but I couldn't get my head around what would happen if I was doing subtractions to unsigned types - if I have to do unsigned types a diffetent way then I've not made anything any simpler. So I can't think of a way to simplify the code, but am willing to try anything if you have any other ideas.

Reply:
Ah, I did missed that.

But what if we put the `if` branch only for `checked_add` and `checked_sub` but keep the `to_digit` and `checked_mul`?

Can you also try using if it's negative after the for loop to do `checked_neg` there? Remove the for loop and add a branch to convert the positive number to negative using `checked_neg` and return `NegOverflow` if there are issues so we don't need another loop for negative, a single branch.

While trying that, maybe can also try a hand crafted negation instead of `checked_add`? Like this
```rust
    for &c in digits {
        ... // checked_add ...
    }
    if !is_positive {
        const UNSIGNED_BITS: i8 = (i8::MAX as u8 * 2) as i8 + 1; // don't just use i8, can use whatever type it is
        result = (i ^ UNSIGNED_BITS).checked_add(1);
    }
```

One more thing, I don't think we need the whole bunch of negative calculation in unsigned code do we?

Reply:
For `i8` if we're parsing `"-128"` we run into representation trouble if we start positive and then negate, that's why one would want to do subtractions and then neg. But we'd need a different code path for unsigned...

>  I don't think we need the whole bunch of negative calculation in unsigned code do we?

Hmm, we could do ` if is_positive || !is_signed_ty {` rather than `if is_positive` - that way the compiler should remove the else block for unsigned types (I don't _think_ currently the compiler can realise that is_positive is always true for unsigned types, but it's pretty crafty so it _might_ have figured that out anyhows).

Reply:
I'll fork your parseint and add some benchmarks.  (The `trick` is very cool you've got there btw.)

I need to double check my benchmarks anyhow because I reaslised that my blackbox was not quite in the right place and the compiler was optimising for the data (rookie mistake!).

Reply:
You could move that for loop into a closure or macro or something that gets `PosOverflow`/`NegOverflow` and `checked_add`/`checked_sub` as parameters.

Reply:
Digging a little with godbolt it can't evaluate at compile time `let is_signed_ty = T::from_u32(0) > T::min_value();` so it doesn't realise it doesn't need the negative side of the if. 

But I'm sure we can do this. `T::min_value()` can be replaced with `T::MIN` as assocated constants are allowed on traits these days (yay!). We can almost use `T::defult()` to get the number zero but it's not const. We could define a const called ZERO on our integers (num crate's zero trait says it can't do this because of bignums). T::MIN + T::MAX always equals -1 but I don't think that helps... is there a cunning way of creating zero? If not passing is_signed as a macro parameter would work...

Reply:
Could you link your godbolt experiments?

Reply:
Sure, on the left it's a const that evaluates to false and on the right it's an expression.  https://godbolt.org/z/od3ajTh3v
Interestingly if I replace 
   `const is_signed_ty: bool = false;` with 
   `   let is_signed_ty: bool = false;` 
it can't tell that it should drop the else expression which is surprising. I would have thought it should have promoted the bool to being a constant given it was assigned by a constant and is not mutable.

Reply:
That godbolt link shows optimized assembly. You forgot to set `-Copt-level=3`. With that the assembly is identical between the two.

Reply:
Or `-O`

Reply:
@pickfire I took a stab at building in all the safeguards to your trick as an alternative route for radix 10. Have submitted a PR for us to talk around: https://github.com/pickfire/parseint/pull/2